File: client/typescript/src/cliUtils.ts
import { env } from "node:process";

export const DEFAULT_URL = env.LOOMI_MCP_URL || "https://mcp.bloomreach.com/mcp";

export interface RawArgs {
  url: string;
  q: string | null;
  rows: number;
  facet: boolean;
  extra: Record<string, string | string[]>;
}

export function parseArgs(raw: string[]): RawArgs {
  const args: RawArgs = { url: DEFAULT_URL, q: null, rows: 12, facet: false, extra: {} };
  for (let i = 2; i < raw.length; i++) {
    const token = raw[i];
    if (token === "--url") {
      args.url = raw[++i] || args.url;
    } else if (token === "--q") {
      args.q = raw[++i] || null;
    } else if (token === "--rows") {
      const n = Number(raw[++i]);
      if (!Number.isFinite(n) || n < 0) throw new Error("--rows must be a non-negative number");
      args.rows = n;
    } else if (token === "--facet") {
      args.facet = true;
    } else if (token === "--arg") {
      const pair = raw[++i];
      if (!pair || !pair.includes("=")) throw new Error("Invalid --arg format, expected KEY=VALUE");
      const [k, vRaw] = pair.split("=", 2);
      const kTrim = k.trim();
      const vTrim = vRaw.trim();
      if (vTrim.includes(",")) {
        args.extra[kTrim] = vTrim.split(",").map((s) => s.trim());
      } else {
        args.extra[kTrim] = vTrim;
      }
    }
  }
  return args;
}

export type HeadersMap = Record<string, string>;

export function buildHeaders(): HeadersMap {
  const headers: HeadersMap = {};
  const accountId = env.ACCOUNT_ID || env.LOOMI_ACCOUNT_ID;
  const domainKey = env.DOMAIN_KEY || env.LOOMI_DOMAIN_KEY;
  const authKey = env.AUTH_KEY || env.LOOMI_AUTH_KEY;
  if (accountId) headers["account_id"] = accountId;
  if (domainKey) headers["domain_key"] = domainKey;
  if (authKey) headers["auth_key"] = authKey;
  return headers;
}

export interface SearchParamsArguments {
  q: string;
  rows: number;
  facet: string;
  [key: string]: string | number | string[];
}

export function prepareArguments({ q, rows, facet, extra }: RawArgs): SearchParamsArguments {
  if (!q || q.trim() === "*" || q.trim().length === 0) {
    throw new Error("q must be non-empty and not equal to '*'");
  }
  const args: SearchParamsArguments = { q: q.trim(), rows, facet: facet ? "true" : "false" };
  if (extra && typeof extra === "object") {
    const merged: Record<string, string | string[]> = { ...extra };
    if (typeof (merged as any).facet === "boolean") {
      (merged as any).facet = (merged as any).facet ? "true" : "false";
    }
    Object.assign(args, merged);
  }
  return args;
}

File: client/python/lib/cli_utils.py
from __future__ import annotations

import argparse
import os
from typing import Any, Dict, Optional

DEFAULT_URL = os.getenv("LOOMI_MCP_URL", "https://mcp.bloomreach.com/mcp")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Python MCP SSE client for Bloomreach Discovery search"
    )
    parser.add_argument(
        "--url",
        default=DEFAULT_URL,
        help="MCP endpoint base URL (e.g., https://mcp.bloomreach.com/mcp).",
    )
    parser.add_argument(
        "--account-id",
        default=os.getenv("ACCOUNT_ID") or os.getenv("LOOMI_ACCOUNT_ID"),
        help="Bloomreach account id (env: ACCOUNT_ID)",
    )
    parser.add_argument(
        "--domain-key",
        default=os.getenv("DOMAIN_KEY") or os.getenv("LOOMI_DOMAIN_KEY"),
        help="Bloomreach domain key (env: DOMAIN_KEY)",
    )
    parser.add_argument(
        "--auth-key",
        default=os.getenv("AUTH_KEY") or os.getenv("LOOMI_AUTH_KEY"),
        help="Bloomreach auth key (optional unless server requires; env: AUTH_KEY)",
    )
    parser.add_argument(
        "--q",
        required=True,
        help="Search query string (must not be '*')",
    )
    parser.add_argument(
        "--rows",
        type=int,
        default=12,
        help="Number of rows to return (default: %(default)s)",
    )
    parser.add_argument(
        "--facet",
        action="store_true",
        help="Enable faceting (default: disabled)",
    )
    parser.add_argument(
        "--arg",
        action="append",
        metavar="KEY=VALUE",
        help="Additional search argument(s), e.g. --arg fq=category:chair. Can be repeated.",
    )
    return parser.parse_args()


def build_headers(
    account_id: Optional[str],
    domain_key: Optional[str],
    auth_key: Optional[str],
) -> Dict[str, str]:
    headers: Dict[str, str] = {}
    if account_id:
        headers["account_id"] = account_id
    if domain_key:
        headers["domain_key"] = domain_key
    if auth_key:
        headers["auth_key"] = auth_key
    return headers


def parse_key_value_pairs(pairs: Optional[list[str]]) -> Dict[str, Any]:
    if not pairs:
        return {}
    parsed: Dict[str, Any] = {}
    for item in pairs:
        if "=" not in item:
            raise ValueError(f"Invalid --arg format, expected KEY=VALUE, got: {item}")
        key, value = item.split("=", 1)
        key = key.strip()
        value = value.strip()
        if "," in value:
            parsed[key] = [v.strip() for v in value.split(",")]
        else:
            parsed[key] = value
    return parsed


def prepare_search_arguments(
    query: str,
    rows: int,
    facet: bool,
    extra_args: Optional[Dict[str, Any]],
) -> Dict[str, Any]:
    if not query or query.strip() == "*":
        raise ValueError("Query q must be a non-empty value other than '*'")
    arguments: Dict[str, Any] = {
        "q": query.strip(),
        "rows": rows,
        "facet": "true" if facet else "false",
    }
    if extra_args:
        merged = dict(extra_args)
        if "facet" in merged and isinstance(merged["facet"], bool):
            merged["facet"] = "true" if merged["facet"] else "false"
        arguments.update(merged)
    return arguments


